### ISSUE TITLE
Problems building v8.10 branch

### ISSUE THREAD

**Issue by palmskog** [ejgallego/coq-serapi#143]
I'm using Coq (`8.10.dev`) and OCaml packages via OPAM, and it doesn't work to build the `v8.10` branch. The problems appear to be twofold:

Coq plugin linking not working (due to some build/install issue with the OPAM package?):
```
File "serlib/plugins/firstorder/dune", line 7, characters 12-34:
7 |  (libraries coq.plugins.firstorder serlib sexplib))
                ^^^^^^^^^^^^^^^^^^^^^^
Error: Library "coq.plugins.firstorder" not found.
Hint: try: dune external-lib-deps --missing @@default
File "serlib/plugins/setoid_ring/dune", line 7, characters 12-35:
7 |  (libraries coq.plugins.setoid_ring serlib serlib_ltac sexplib))
                ^^^^^^^^^^^^^^^^^^^^^^^
Error: Library "coq.plugins.setoid_ring" not found.
```

Yojson library issues:
```
File "serlib/serType.ml", line 15, characters 18-31:
Error: Unbound type constructor Yojson.Safe.t

File "serlib/ser_context.mli", line 23, characters 30-43:
Error: Unbound type constructor Yojson.Safe.t

File "serlib/ser_constr.mli", line 107, characters 16-29:
Error: Unbound type constructor Yojson.Safe.t
``` 

The OPAM package `yojson` is at version 1.5.0, which is due to `ppx_deriving_yojson` (3.3) requiring `< 1.6.0`.

**Comment by ejgallego** [comment]
Ack. The serapi build expects a dune-built Coq, I'll make a fix upstream so the make based one does work (meanwhile pinning to be branch should work)

As for yojson indeed, sorry. This is an upstream problem who messed up the yojson update (see thread on discuss.ocaml) We either revert the Json patch or you'll have to pin ppx_deriving_yojson to their dev branch.

**Comment by ejgallego** [comment]
https://github.com/coq/coq/pull/10009 should fix the first problem, as for the second, we can only for a `ppx_deriving_yojson`release soon c.f. https://github.com/ocaml-ppx/ppx_deriving_yojson/issues/94

**Comment by ejgallego** [comment]
There are more problems related to broken packing in the default 8.10 build; if you pin the branch the Dune build will take over and it works.

We can try to add a workaround but it may take a while.

**Comment by palmskog** [comment]
I'm fine with with pinning the branch for now, but for less advanced users I think this should be solved in some more robust way at some reasonable point in the future, e.g., by changing the OPAM package in the Coq OPAM archive.

**Comment by palmskog** [comment]
Also, I think we should document a recommended non-standard build very clearly, ideally in `README.md`.

**Comment by ejgallego** [comment]
> Also, I think we should document a recommended non-standard build very clearly, ideally in `README.md`.

For "dev" branches I let the Travis file to take that role, as long as it is green it should be up to date; but indeed, we need to fix this before release.

Something that was not discussed is if the official Coq OPAM package will use Dune [which is what one in-repos] does.

We also have a quite nice "dev build", however it depends on an unreleased version of Dune so I didn't push it to the repos yet [you do `make dev` and all magic happens]

**Comment by ejgallego** [comment]
The problem with the non-Dune build is very painful as it turns out that the interfaces for `Ltac_plugin.Foo` are not in its proper place.

Indeed we can workaround that but it will be painful.

**Comment by palmskog** [comment]
Speaking of builds (commands in `.travis.yml`), we should probably not depend on mathcomp, I think the only use is in in `tests/genarg/move.v`.

**Comment by ejgallego** [comment]
> Speaking of builds (commands in `.travis.yml`), we should probably not depend on mathcomp, I think the only use is in in `tests/genarg/move.v`.

Oh, you think so @palmskog ? I was happy indeed that we were testing math-comp support for the serialization plugin.

**Comment by palmskog** [comment]
Yes, but we can do that without actually requiring anything from MathComp itself, the stuff in `plugins/ssr` should be enough.

**Comment by ejgallego** [comment]
Ah, I see what you mean, sure; let's cleanup that when we can.

Tho, it doesn't bother me to depend on somehow "mature" libraries for the test suite if needed.

**Comment by ejgallego** [comment]
Actually I am not sure how to fix things so ppx_import works with both packing styles :(

It seems to me that the best path would be to use the Dune build in the opam package indeed.

**Comment by palmskog** [comment]
@ejgallego can you bottom-line this one for me if I want to continue to install Coq via OPAM and compile SerAPI `v8.10`? Should I submit a Dune-based OPAM package for Coq's `v8.10` to `core-dev`, or do something else?

**Comment by palmskog** [comment]
And speaking of Dune and builds, I'm fairly certain that Dune 1.2 is not going to cut it anymore as per SerAPI's `dune-project`, in particular I believe the `fmt` extension is not available for that Dune version. Might be a good idea to bump.

**Comment by ejgallego** [comment]
The short summary is that I didn't figure out yet how to make the plugins serializers work with both build systems :( Each attempt I tried fails, thus I dunno what to do. I guess I need to think a bit more but the fix is looking really involved :( If you have an idea how to fix this problem people don't hesistate.

This is very unfortunate as I crucially rely on the Dune-composed build to develop.

I won't be able to think about this in the next few days. 

I am not sure about OPAM, your suggestion would certainly fix it. Maybe we could even make the main opam package to be built with dune, IMO it is in better shape, but YMMV.

> Might be a good idea to bump.

Sure thanks, I'll bump at least to the Coq minimum [feel free to bump it yourself]

**Comment by ejgallego** [comment]
This is almost solved by #179 , still the same codebase is not compatible with a make-built 8.10 and a dune-built 8.11, however the diff is down to a very few lines; I have a workaround in mind.
